pkg/client/client.go

```diff
@@ -0,0 +1,410 @@
+	// be set to the provided new value.
+	OnConfig(string)
+
+	// OnStop is called when the Elastic Agent is requesting the application to stop.
+	OnStop()
+
+	// OnError is called when an errors occurs communicating with Elastic Agent.
+	//
+	// These error messages are not given by the Elastic Agent, they are just errors exposed
+	// from the client-side GRPC connection.
+	OnError(error)
+}
+
+// Client manages the state and communication to the Elastic Agent.
+type Client struct {
+	target          string
+	opts            []grpc.DialOption
+	token           string
+	impl            StateInterface
+	actions         map[string]Action
+	cfgIdx          uint64
+	cfg             string
+	expected        proto.StateExpected_State
+	observed        proto.StateObserved_Status
+	observedMessage string
+
+	ctx    context.Context
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+	client proto.ElasticAgentClient
+	cfgLock   sync.RWMutex
+	obsLock   sync.RWMutex
+
+	// overridden in tests to make fast
+	minCheckTimeout time.Duration
+}
+
+// New creates a client connection to Elastic Agent.
+func New(target string, token string, impl StateInterface, actions []Action, opts ...grpc.DialOption) *Client {
+	actionMap := map[string]Action{}
+	if actions != nil {
+		for _, act := range actions {
+			actionMap[act.Name()] = act
+		}
+	}
+	return &Client{
+		target:          target,
+		opts:            opts,
+		token:           token,
+		impl:            impl,
+		actions:         actionMap,
+		cfgIdx:          0, // 0 is initial, no config state
+		expected:        proto.StateExpected_RUNNING,
+		observed:        proto.StateObserved_STARTING,
+		observedMessage: "Starting",
+		minCheckTimeout: CheckinMinimumTimeout,
+	}
+}
+
+// Start starts the connection to Elastic Agent.
+func (c *Client) Start(ctx context.Context) error {
+	ctx, cancel := context.WithCancel(ctx)
+	c.ctx = ctx
+	c.cancel = cancel
+	conn, err := grpc.DialContext(ctx, c.target, c.opts...)
+	if err != nil {
+		return err
+	}
+	c.client = proto.NewElasticAgentClient(conn)
+	c.startCheckin()
+	c.startActions()
+	return nil
+}
+
+// Stop stops the connection to Elastic Agent.
+func (c *Client) Stop() {
+	if c.cancel != nil {
+		c.cancel()
+		c.wg.Wait()
+		c.ctx = nil
+		c.cancel = nil
+	}
+}
+
+// Status updates the current status of the client in the Elastic Agent.
+func (c *Client) Status(status proto.StateObserved_Status, message string) {
+	c.obsLock.Lock()
+	c.observed = status
+	c.observedMessage = message
+	c.obsLock.Unlock()
+}
+
+// startCheckin starts the go rountines to send and receive check-ins
+func (c *Client) startCheckin() {
+	c.wg.Add(1)
+
+	go func() {
+		defer c.wg.Done()
+		for {
+			select {
+			case <-c.ctx.Done():
+				// stopped
+				return
+			default:
+			}
+
+			checkinClient, err := c.client.Checkin(c.ctx)
+			if err != nil {
+				c.impl.OnError(err)
+				continue
+			}
+
+			var wg sync.WaitGroup
+			wg.Add(2)
```

Comment:
When I was reading this my initial question was: Why `2`. I then found out further down the second wg.Done. I wonder if we should have `wg.Add(1)` on line 176 and the second one line 211. I think it would make the code more readable to have Add and Done closer together.

Reply:
I have moved them closer.